inspirehep/modules/records/permissions.py

```diff
@@ -0,0 +1,148 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2016 CERN.
+#
+# INSPIRE is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this licence, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+from __future__ import absolute_import, division, print_function
+
+import pickle
+
+from flask import current_app, session
+from flask_security import current_user
+from werkzeug.local import LocalProxy
+
+from invenio_access.models import ActionUsers
+from invenio_access.permissions import (DynamicPermission,
+                                        ParameterizedActionNeed)
+
+from inspirehep.utils.cache import cache
+
+
+action_view_restricted_collection = ParameterizedActionNeed(
+    'view-restricted-collection', argument=None
+)
+
+all_restricted_collections = LocalProxy(lambda: load_restricted_collections())
+
+user_collections = LocalProxy(lambda: get_user_collections())
+
+
+def get_user_collections():
+    """Get user restricted collections."""
+    return session.get('restricted_collections', set())
+
+
+def load_user_collections(app, user):
+    """Load user restricted collections upon login.
+
+    Receiver for flask_login.user_logged_in
+    """
+    user_collections = set(
+        [a.argument for a in ActionUsers.query.filter_by(
+            action='view-restricted-collection',
+            user_id=current_user.get_id()).all()]
+    )
+    session['restricted_collections'] = user_collections
+
+
+def load_restricted_collections():
+    """Load and cache all restricted collections."""
+    restricted_collections = cache.get('restricted_collections')
+    if restricted_collections:
+        return pickle.loads(restricted_collections)
+    else:
+        restricted_collections = set(
+            [
+                a.argument for a in ActionUsers.query.filter_by(
+                    action='view-restricted-collection').all()
+            ]
+        )
+        if restricted_collections:
+            cache.set(
+                'restricted_collections',
+                pickle.dumps(restricted_collections),
```

Comment:
I will use https://github.com/inspirehep/inspire-next/pull/1622 so will not do pickling in the end myself.